### DISCUSSION TITLE
Plex Rich Presence 2.0

### DISCUSSION THREAD

**Discussion by Ombrelin** [Ombrelin/plex-rich-presence#66] [Q&A] [answered]
Hello everyone,

I'm starting to plan the project of remaking the Plex Rich Presence app. I'm making this post to announce the planned features, which were suggested by the users, but also because I still have some things I still haven't decided.

## Tech Stack

- Runtime & Language : .NET 6 & C# 10
- Library for discord rich presence : [Discord Game SDK](https://discord.com/developers/docs/game-sdk/sdk-starter-guide)
- Library for PLEX :  [PLEX API](https://github.com/jensenkd/plex-api)

The goal will be to have a code base which is tested and of quality (leveraging IoC, Clean Code...)

## Planned features

- Ability to choose one server among the servers available on the PLEX account (#9)
- Support costumization (with an interactive UI) of what is displayed in the rich presence caption (#60, #43 )
- Support displaying image to put track / album / artist image in the rich presence (thanks to Discord Game SDK) (#55 )
- Usage of PLEX Websockets API to enable realtime (instead of polling) session status and allow any user of the server (not just owner) to use the app (#40, #68 )
- Optionaly choose device to track activity from (#59, #6 )
- CLI Version (#56)

## Planned bug fixes

- Rich presence bugger when no track id in meta-data (#58 )
- Handle loss of connection to the server (#63 )

## Questioning

I'm strongly hesitating on one point of the tech stack : Which UI framework to use ?
.NET has two frameworks for cross-plateform UI development : [Avalonia](https://github.com/AvaloniaUI/Avalonia) and [.NET MAUI](https://github.com/dotnet/maui).

Each have their advantages and drawbacks, what do you think ?

### Avalonia 

- Fairly mature
- Less polished Tray Icon API
- Cross platform across Windows, Mac and Linux
- Not an official framework
- Constant UI across platform (drawn controls using Skia)

### MAUI

- Not in GA yet
- More polished Tray Icon API
- Cross platform across Windows and Mac for desktop, no Linux
- Official framework maintained by Microsoft
- Use platform-native UI, mapping to native controls

**Comment by darovic** [comment]
MAUI looks like it might be the way to go. Allegedly, there is Linux support provided by the community, but I can't see any links (source: https://docs.microsoft.com/en-us/dotnet/maui/supported-platforms). I'll do some research and get back to you.

**Reply by Ombrelin** [reply]
Yes you're right, I should have written "no official linux support from microsoft". The current community project for Linux support (leveraging GtkSharp), is the following : https://github.com/jsuarezruiz/maui-linux) but the state of the "Status" page of the wiki does not give any info on whether it's usable at the time.

**Comment by itslino** [comment]
I'd love to be able to display art of music that I'm listening to, there's another project that got images to load up for TV Shows and Movies using a TMDB api key.

https://github.com/EthanC/Perplex

The only thing I noticed is that the sessions crashed but its probably cause how early its in development. But maybe that could be one approach?

**Reply by Ombrelin** [reply]
Hello,
As listed with #55 I plan on displaying poster of the media by leveraging the discord game SDK.
I planned to fetch the image from plex directly since plex manages meta data including poster, and leverages TMDB in some cases.
I don't uite understand the added value in fetching art directly from TMDB in the app ; could you elaborate on that matter ?
Thank you in advance,

**Reply by itslino** [reply]
I'm not sure honestly, I do know that someone made a mod where you can fetch music album art from Tatulli but it wouldn't work if the images were pulled locally in my Tautulli. You'd need a public Tautulli otherwise it showed the corrupted image file.

Maybe this is why he chose to use to source images from a public web source??

You can see the comments of those who modified the existing project on here.
https://pythonrepo.com/repo/EthanC-Perplex

**Reply by itslino** [reply]
Here's the dev's explanation
https://github.com/EthanC/Perplex/pull/4

**Reply by Ombrelin** [reply]
Okay thank you for the material ! I will look into it.
But my first guess is that if I manage to fetch the data directly from the plex server (and not tautulli) which in any case must be accessible from the user computer for the app to work, that will not cause the problem you're talking about.
Anyway I'll think about it and keep you posted about my conslusion, thank you for your interest

**Reply by darovic** [reply]
Given Plex's work integrating Tidal into Plexamp, I suspect letting plex do the heavy lifting will be sufficient for @itslino's needs and saves the hassle of implementing and maintaining a custom scraping strategy.

(And I haven't forgotten to go digging into the linux support for MAUI, I just haven't had the time T.T)

**Reply by Ombrelin** [reply]
Haha no problem we have all the the time we want, no pressure !

**Comment by Jenni32222** [comment]
Great app! Can't wait for V2!

**Comment by stewieboo** [comment]
Really excited! Been using the program since last year. Honestly, I came here to check if there's a way to display posters on my Discord status. Well, it seems like you're already working on it. Mad respect!

**Comment by MargeyShah** [comment]
I'd love to contribute as well! Also, is there a possibility to have Tidal support? I know PMS doesn't report something being played when in use, but wasn't sure if this info could be pulled when using PlexAmp.

**Reply by Ombrelin** [reply]
Hello, 
I don't use Tidal, but IIRC, it has already has been request in issue. I'd love to integrate it, but I would require help from someone who knows and uses Tidal. Maybe they offer an API of some sort we could work with ?

**Reply by MargeyShah** [reply]
I did not catch this message. There is no official API but there are some unofficial ones on GitHub.

**Comment by Ombrelin** [comment]
Hello,
I had time to put some progress in this project, and while I'm not finished yet I'd like to keep you guys up to date on what I've done yet.
First, after dabbling a bit with MAUI, I deemed that its tooling is not yet mature (can't run on Rider yet), and the Linux support is nowhere from ready. I don't want to lock out Linux users (I've myself, recently switched to Ubuntu on my main computer). Hence I selected Avalonia for this project, with [this library](https://github.com/amwx/FluentAvalonia) to feature a nice looking fluent v2 UI.

The UI is well under way, I'm working on implementing the Plex notifications correctly, especially to handle non-owner users.
Here are some screen shots of the current (heavily WIP) UI :

![image](https://user-images.githubusercontent.com/47502331/184507209-15d38f47-7a4b-4075-b514-224338013e67.png)
![image](https://user-images.githubusercontent.com/47502331/184507236-39b86a50-d2e0-42bd-9197-5e61a2473c3b.png)


Hope you like it, I'm counting on releasing v2 in September if everything goes as planned.

**Comment by Ombrelin** [comment]
I'm very close to releasing a 2.0-beta, I'm just checking last details to correclty setup CI, it will be released next week.

**Reply by apoapostolov** [reply]
What we can expect in it?

**Reply by Ombrelin** [reply]
I just updated the roadmap : https://github.com/Ombrelin/plex-rich-presence/wiki/Roadmap

**Comment by Ombrelin** [comment, marked as answer]
The 2.0 version is released !
Hope you will like it, and don't hesitate to give feedback, especially for non server owner users, because I couldn't thoroughly manually test it
